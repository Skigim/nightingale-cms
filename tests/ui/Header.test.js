import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('Header Component', () => {
  let mockOnSettingsClick, mockOnManualSave, Component;

  beforeEach(() => {
    // Reset mocks
    mockOnSettingsClick = jest.fn();
    mockOnManualSave = jest.fn();

    // Mock the UI services
    window.NightingaleUI = {
      Button:
        window.Button ||
        function Button({ onClick, children, className, variant, size }) {
          const e = window.React.createElement;
          const handleClick = (event) => {
            if (onClick) {
              onClick(event);
            }
          };
          return e(
            'button',
            {
              onClick: handleClick,
              className: `btn ${variant || ''} ${size || ''} ${className || ''}`,
            },
            children
          );
        },
    };

    // Mock the Header component based on actual API
    Component = function Header({
      filename,
      fileStatus,
      autosaveStatus,
      onSettingsClick,
      onManualSave,
    }) {
      const e = window.React.createElement;

      // Only show manual save button when autosave fails or is missing
      const showManualSave =
        !autosaveStatus ||
        autosaveStatus.status === 'error' ||
        autosaveStatus.status === 'stopped' ||
        !autosaveStatus.status;

      return e(
        'header',
        { className: 'header bg-gray-900 text-white p-4' },
        e(
          'div',
          { className: 'flex justify-between items-center' },
          e(
            'div',
            { className: 'flex items-center gap-4' },
            e('h1', { className: 'text-xl font-bold' }, 'Nightingale CMS'),
            filename && e('span', { className: 'text-gray-300' }, filename),
            fileStatus &&
              e(
                'span',
                {
                  className: `status-indicator ${fileStatus === 'saved' ? 'text-green-400' : 'text-yellow-400'}`,
                },
                fileStatus
              )
          ),
          e(
            'div',
            { className: 'flex items-center gap-2' },
            autosaveStatus &&
              e(
                'div',
                { className: 'autosave-status' },
                e(
                  'span',
                  { className: 'text-sm text-gray-400' },
                  `${autosaveStatus.status || 'unknown'}: ${autosaveStatus.message || 'No message'}`
                )
              ),
            showManualSave &&
              e(
                window.NightingaleUI.Button,
                {
                  onClick: onManualSave,
                  variant: 'primary',
                  size: 'sm',
                  className: 'save-button',
                },
                'Save'
              ),
            e(
              window.NightingaleUI.Button,
              {
                onClick: onSettingsClick,
                variant: 'secondary',
                size: 'sm',
                className: 'settings-button',
              },
              'Settings'
            )
          )
        )
      );
    };

    // Make component available globally
    window.Header = Component;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const defaultProps = {
    filename: 'test.json',
    fileStatus: 'saved',
    autosaveStatus: { status: 'connected', message: 'Ready' },
    onSettingsClick: mockOnSettingsClick,
    onManualSave: mockOnManualSave,
  };

  test('renders header with title', () => {
    render(React.createElement(Component, defaultProps));

    expect(screen.getByText('Nightingale CMS')).toBeInTheDocument();
  });

  test('displays filename when provided', () => {
    render(React.createElement(Component, defaultProps));

    expect(screen.getByText('test.json')).toBeInTheDocument();
  });

  test('displays file status when provided', () => {
    render(React.createElement(Component, defaultProps));

    expect(screen.getByText('saved')).toBeInTheDocument();
  });

  test('shows autosave status when provided', () => {
    render(React.createElement(Component, defaultProps));

    expect(screen.getByText('connected: Ready')).toBeInTheDocument();
  });

  test('calls onSettingsClick when settings button is clicked', () => {
    render(React.createElement(Component, defaultProps));

    const settingsButton = screen.getByText('Settings');
    fireEvent.click(settingsButton);

    expect(mockOnSettingsClick).toHaveBeenCalled();
  });

  test('shows save button only when autosave status is error', () => {
    const props = {
      ...defaultProps,
      autosaveStatus: { status: 'error', message: 'Failed to save' },
    };

    render(React.createElement(Component, props));

    expect(screen.getByText('Save')).toBeInTheDocument();
  });

  test('shows save button when autosave status is stopped', () => {
    const props = {
      ...defaultProps,
      autosaveStatus: { status: 'stopped', message: 'Autosave disabled' },
    };

    render(React.createElement(Component, props));

    expect(screen.getByText('Save')).toBeInTheDocument();
  });

  test('hides save button when autosave is working normally', () => {
    const props = {
      ...defaultProps,
      autosaveStatus: { status: 'connected', message: 'Ready' },
    };

    render(React.createElement(Component, props));

    expect(screen.queryByText('Save')).not.toBeInTheDocument();
  });

  test('calls onManualSave when save button is clicked', () => {
    const props = {
      ...defaultProps,
      autosaveStatus: { status: 'error', message: 'Failed to save' },
    };

    render(React.createElement(Component, props));

    const saveButton = screen.getByText('Save');
    fireEvent.click(saveButton);

    expect(mockOnManualSave).toHaveBeenCalled();
  });

  test('handles missing filename gracefully', () => {
    const props = { ...defaultProps, filename: null };

    render(React.createElement(Component, props));

    expect(screen.getByText('Nightingale CMS')).toBeInTheDocument();
    expect(screen.queryByText('test.json')).not.toBeInTheDocument();
  });

  test('handles missing file status gracefully', () => {
    const props = { ...defaultProps, fileStatus: null };

    render(React.createElement(Component, props));

    expect(screen.getByText('Nightingale CMS')).toBeInTheDocument();
    expect(screen.queryByText('saved')).not.toBeInTheDocument();
  });

  test('handles missing autosave status gracefully', () => {
    const props = { ...defaultProps, autosaveStatus: null };

    render(React.createElement(Component, props));

    expect(screen.getByText('Nightingale CMS')).toBeInTheDocument();
    // When autosave status is missing, save button should show
    expect(screen.getByText('Save')).toBeInTheDocument();
  });

  test('renders without crashing with minimal props', () => {
    const minimalProps = {
      onSettingsClick: mockOnSettingsClick,
      onManualSave: mockOnManualSave,
    };

    render(React.createElement(Component, minimalProps));

    expect(screen.getByText('Nightingale CMS')).toBeInTheDocument();
    expect(screen.getByText('Settings')).toBeInTheDocument();
  });

  test('applies correct CSS classes', () => {
    const { container } = render(React.createElement(Component, defaultProps));

    const header = container.querySelector('header');
    expect(header).toHaveClass('header', 'bg-gray-900', 'text-white', 'p-4');
  });

  test('maintains component structure', () => {
    const { container } = render(React.createElement(Component, defaultProps));

    expect(container.querySelector('header')).toBeInTheDocument();
    expect(
      container.querySelector('.flex.justify-between')
    ).toBeInTheDocument();
    expect(container.querySelector('h1')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { container } = render(React.createElement(Component, defaultProps));

    expect(container.firstChild).toMatchSnapshot();
  });
});
