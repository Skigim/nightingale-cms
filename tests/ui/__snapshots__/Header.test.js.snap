// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component matches snapshot 1`] = `
<div
  class="flex items-center justify-between bg-gray-800 p-2 shadow-md flex-shrink-0"
>
  <div
    class="flex items-center"
  >
    <svg
      class="h-6 w-6 mr-2 text-blue-400"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 12h6m-6 4h6m2 5H7a2 2 0 01-2-2V5a2 2 0 012-2h5.586a1 1 0 01.707.293l5.414 5.414a1 1 0 01.293.707V19a2 2 0 01-2 2z"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
    <span
      class="font-bold text-white"
    >
      Nightingale CMS - React
    </span>
  </div>
  <div
    class="flex items-center space-x-3"
  >
    <div
      class="text-sm"
    >
      <button
        class="px-2 py-1 rounded text-xs font-medium transition-colors flex items-center space-x-1 bg-green-600 text-green-100 hover:bg-green-500 cursor-pointer"
        title="Click to open settings and change save folder"
      >
        <span>
          Connected
        </span>
        <svg
          class="w-3 h-3"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M10.325 4.317c.426-1.756 2.924-1.756 3.35 0a1.724 1.724 0 002.573 1.066c1.543-.94 3.31.826 2.37 2.37a1.724 1.724 0 001.065 2.572c1.756.426 1.756 2.924 0 3.35a1.724 1.724 0 00-1.066 2.573c.94 1.543-.826 3.31-2.37 2.37a1.724 1.724 0 00-2.572 1.065c-.426 1.756-2.924 1.756-3.35 0a1.724 1.724 0 00-2.573-1.066c-1.543.94-3.31-.826-2.37-2.37a1.724 1.724 0 00-1.065-2.572c-1.756-.426-1.756-2.924 0-3.35a1.724 1.724 0 001.066-2.573c-.94-1.543.826-3.31 2.37-2.37.996.608 2.296.096 2.572-1.065z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
    <div
      class="flex items-center space-x-1.5"
      title="Auto-saved 2 minutes ago"
    >
      <span
        class="text-sm"
      >
        âœ“
      </span>
      <span
        class="text-xs text-green-400"
      >
        Saved
      </span>
    </div>
    <div
      class="flex space-x-2"
    >
      <div
        class="w-3 h-3 bg-yellow-400 rounded-full"
      />
      <div
        class="w-3 h-3 bg-green-400 rounded-full"
      />
      <div
        class="w-3 h-3 bg-red-400 rounded-full"
      />
    </div>
  </div>
</div>
`;
